### PR TITLE
 fix(TBD-11038): The distribution AMAZON_EMR with the version APACHE_2_4_0_EMR doesn't exist

### PR DESCRIPTION
**Pull request by cbadillo1603**
…2_4_0_EMR doesn't exist

**What is the current behavior?** (You can also link to an open issue here)

**What is the new behavior?**

**Please check if the PR fulfills these requirements**

- [ ] The commit message follows Talend standard
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features) ?
- [ ] The code coverage on new code >75%
- [ ] The new code does not introduce new technical issues (sonar / eslint)

**What kind of change does this PR introduce?**

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build / CI related changes
- [ ] Other... Please describe:

**Does this PR introduce a breaking change?**

- [ ] Yes
- [ ] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

File: main/plugins/org.talend.repository/src/main/java/org/talend/repository/model/migration/spark/UpdateRemovedHadoopVersion.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    
+    private ComponentType componentType = ComponentType.getComponentType("HDFS");
+    private DistributionsManager distributionsHelper = new DistributionsManager(componentType);
+    private List<DistributionBean> distros = Arrays.asList( distributionsHelper.getDistributions() );
+   
+    private List<DistributionVersion> versions = distros.stream().map( d -> Arrays.asList( d.getVersions() ) )
+                                                        .flatMap(List::stream)
+                                                        .collect(Collectors.toList());
+   
+    private List<String> versionsLabel = versions.stream().map( v -> v.version ).collect(Collectors.toList());
```
Comment (amedvedenko) [inline main/plugins/org.talend.repository/src/main/java/org/talend/repository/model/migration/spark/UpdateRemovedHadoopVersion.java:59]: this logic should be in constructor

File: main/plugins/org.talend.repository/src/main/java/org/talend/repository/model/migration/spark/UpdateRemovedHadoopVersion.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    
+    
+    @Override
+    public List<ERepositoryObjectType> getTypes() {
+        List<ERepositoryObjectType> toReturn = new ArrayList<ERepositoryObjectType>();
```
Comment (amedvedenko) [inline main/plugins/org.talend.repository/src/main/java/org/talend/repository/model/migration/spark/UpdateRemovedHadoopVersion.java:64]: if you use lamda expressions then diamond operators should also be available)

File: main/plugins/org.talend.repository/src/main/java/org/talend/repository/model/migration/spark/UpdateRemovedHadoopVersion.java (hunk 1 above)
Comment (amedvedenko) [reply main/plugins/org.talend.repository/src/main/java/org/talend/repository/model/migration/spark/UpdateRemovedHadoopVersion.java:59]: I think everything except IMPACTED_COMPONENT_TYPES should be initialized in constructor

File: main/plugins/org.talend.repository/src/main/java/org/talend/repository/model/migration/spark/UpdateRemovedHadoopVersion.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+import org.talend.designer.core.model.utils.emf.talendfile.ElementParameterType;
+
+/**
+ * Migration task to update the value of spark version when breaking components.
+ * If its a spark local job, the distrib version should not affect the job
+ * if its not a spark local job, the spark local version should not affect the job
+ */
+public class UpdateRemovedHadoopVersion extends AbstractJobMigrationTask {
+    
+    private final List<String> IMPACTED_COMPONENT_TYPES =
```
Comment (amedvedenko) [inline main/plugins/org.talend.repository/src/main/java/org/talend/repository/model/migration/spark/UpdateRemovedHadoopVersion.java:47]: it isn't static but you use UPPER_CASE style, please rename it to lowerCase or make it static

File: main/plugins/org.talend.repository/src/main/java/org/talend/repository/model/migration/spark/UpdateRemovedHadoopVersion.java (hunk 1 above)
Comment (amedvedenko) [reply main/plugins/org.talend.repository/src/main/java/org/talend/repository/model/migration/spark/UpdateRemovedHadoopVersion.java:59]: and all variables can be final, so I suggest adding final keyword to all class-level vars where it's possible.

File: main/plugins/org.talend.repository/src/main/java/org/talend/repository/model/migration/spark/UpdateRemovedHadoopVersion.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        GregorianCalendar gc = new GregorianCalendar(2020, 10, 05, 13, 0, 0);
+        return gc.getTime();
+    }
+    
+    private class HadoopVersionCoverter implements IComponentConversion {
+
+        String DEFAULT_DISTRIBUTION = "AMAZON_EMR";
+        String DEFAULT_VERSION = versionsLabel.get(0);
+        
+        public void transform(NodeType node) {
```
Comment (amedvedenko) [inline main/plugins/org.talend.repository/src/main/java/org/talend/repository/model/migration/spark/UpdateRemovedHadoopVersion.java:114]: Is @ Override annotation missed?

File: main/plugins/org.talend.repository/src/main/java/org/talend/repository/model/migration/spark/UpdateRemovedHadoopVersion.java [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    @Override
+    public Date getOrder() {
+        GregorianCalendar gc = new GregorianCalendar(2020, 10, 05, 13, 0, 0);
+        return gc.getTime();
+    }
+    
+    private class HadoopVersionCoverter implements IComponentConversion {
+
+        String DEFAULT_DISTRIBUTION = "AMAZON_EMR";
+        String DEFAULT_VERSION = versionsLabel.get(0);
```
Comment (amedvedenko) [inline main/plugins/org.talend.repository/src/main/java/org/talend/repository/model/migration/spark/UpdateRemovedHadoopVersion.java:112]: .get(0) seems to be not safe, I suggest moving it to constructor. 
Also, what's happened with variable names? :D

File: main/plugins/org.talend.repository/src/main/java/org/talend/repository/model/migration/spark/UpdateRemovedHadoopVersion.java (hunk 3 above)
Comment (cbadillo1603) [reply main/plugins/org.talend.repository/src/main/java/org/talend/repository/model/migration/spark/UpdateRemovedHadoopVersion.java:47]: uppercase names are reserved for final (constant) variables

File: main/plugins/org.talend.repository/src/main/java/org/talend/repository/model/migration/spark/UpdateRemovedHadoopVersion.java (hunk 5 above)
Comment (cbadillo1603) [reply main/plugins/org.talend.repository/src/main/java/org/talend/repository/model/migration/spark/UpdateRemovedHadoopVersion.java]: Even when .get(0) is not safe we need to get a valid distribution from this list. In the remote case where it is empty would mean the whole distribution retrieval module has a problem (Unlikely). So there is no need to invest more effort on making it empty safe.

**Review by amedvedenko — APPROVED** [review]

File: main/plugins/org.talend.repository/META-INF/MANIFEST.MF [hunk 6]
```diff
@@ -51,7 +51,8 @@ Require-Bundle: javax.xml.rpc,
  org.talend.commons.runtime,
  org.talend.designer.maven,
  org.talend.components.api,
- org.eclipse.text
+ org.eclipse.text,
+ org.talend.hadoop.distribution
```
Comment (nrousseau) [inline main/plugins/org.talend.repository/META-INF/MANIFEST.MF:55]: Can't add the hadoop as required here.
BD / hadoop must be optional at this point, and not a required in the base product.

But note that migration tasks can be added in any plugin. (doesn't need to be all in the same..)

**Review by nrousseau — CHANGES_REQUESTED** [review]

File: main/plugins/org.talend.repository/META-INF/MANIFEST.MF (hunk 6 above)
Comment (cbadillo1603) [reply main/plugins/org.talend.repository/META-INF/MANIFEST.MF:55]: Noted, the migration code will be moved to tbd-studio-ee/main/plugins/org.talend.designer.spark/src/org/talend/designer/spark/repository/migration I will close this pull request as is not required any more.

**Comment by cbadillo1603** [comment]
the migration code will be moved to tbd-studio-ee/main/plugins/org.talend.designer.spark/src/org/talend/designer/spark/repository/migration Closing this pull request as is not required any more.

**Pull request closed without merging**